### DISCUSSION TITLE
MyCommand.registerWithSubCommands() when subcommands are inside another module

### DISCUSSION THREAD

**Discussion by Sefrancois** [jmcdo29/nest-commander#817] [Q&A] [answered]
Hello there 👋 

I encountered an issue when trying to work with the method `registerWithSubCommands()` inside my module that exposes my CLI.

What we try to do is to provide & exports a `SubCommand` from one module and imports it to our `CliModule` like following :

```typescript
@SubCommand({
    name: "some-sub-command",
})
export class MySubCommand extends CommandRunner {
    constructor(private readonly someProvider: SomeProvider) {
        super();
    }

    public override run(): void {
        // Do something ...
    }
}
```

Here is how it's instantiated through some random module :

```typescript
@Module({
    imports: [SomeModule],
    providers: [
        SomeProvider,
        {
            provide: MySubCommand,
            inject: [SomeProvider],
            useFactory: (someProvider: SomeProvider): MySubCommand => {
                return new MySubCommand(someProvider);
            }
        }
    ],
    exports: [MySubCommand],
})
export class SomeModule {
}
```

Here we have our command called `MyCommand` :

```typescript
@Command({
    name: "my-command",
    subCommands: [MySubCommand],
})
export class MyCommand extends CommandRunner {
    public override run(): void {
       // Do nothing
    }
}
```

And here is how we want to expose our CLI through our `CliModule` :

```typescript
@Module({
    imports: [SomeModule],
    providers: [...MyCommand.registerWithSubCommands()],
})
export class CliModule {
}
```

Our error when we try to run our application is that Nest can't resolve `MySubCommand` dependencies (`SomeProvider` in that case) as it's not exposed through `SomeModule`. The only thing we export is `MySubCommand`. However, our main problem is more about why does Nest try to instantiate `MySubCommand` again since it's already been instantiated and available through `SomeModule` that we import in our `CliModule` ?

**Comment by jmcdo29** [comment, marked as answer]
Could you get me a repository that has this setup? `nest-commander` shouldn't be trying to make a new instance of that command, but it may be the `moduleRef.get()` that is having issues

**Reply by Sefrancois** [reply]
Hello !

Thanks for your answer !

You'll find some code that reproduces the example in here : https://github.com/Sefrancois/subcommand-issue-example

I joined a screenshot of the error inside the `README.md`

**Reply by jmcdo29** [reply]
Ah! I see. So this actually has to do with how `registerWithSubCommands` works. That static method will find all of the subcommands mentioned in the metadata and add them as providers to the current module. This means that in the `CliModule` you're trying to re-create the subcommands because that's how the `providers` arrays work in NestJS.

If you add `SomeProvider` to the `exports` of your `SomeModule` and remove the `DoSomethingMoreAccurateSubCommand` provider from it, it'll all work as expected. 

I know that might not be ideal for you though

**Reply by Sefrancois** [reply]
Hello there !

Thanks for your answer 😄 

We found out a way to make it work and it was that simple that ... We didn't think of it earlier.

In fact, we don't have to call the method `registerWithSubCommands` since we have already instantiated every single sub command and exported it from `SomeModule`.

So all we have to do in our `CliModule` is the following and it works like a charm :

```typescript
@Module({
    imports: [SomeModule],
    providers: [MyCommand],
})
export class CliModule {
}
```

As `MySubCommand` has already been registered as a provider of `SomeModule` and has been set as an exported provider from `SomeModule`, Nest can instantiate `MyCommand` without trying to instantiate `MySubCommand` again but use existing value.

Just wanted to share it in case somebody else faces the same "issue" 😄
